Fix Ring integration log flooding for accounts without subscription

ptarjan opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

The Ring integration was flooding logs with the warning message "Your Ring account does not have an active subscription" approximately every 3-4 minutes (~480 times daily) for users without a Ring Protect subscription.

The warning originates from the `ring_doorbell` library's `async_recording_url()` method. When the coordinator updates device data, the `live_view` camera entity would attempt to fetch recording URLs from history data, triggering this warning for accounts without subscriptions.

This fix adds a check in `_handle_coordinator_update()` to skip processing recording history for live stream entities when the device doesn't have a subscription. This:
- Prevents log flooding for users without Ring subscriptions
- Preserves existing functionality for users with subscriptions  
- Doesn't affect the live streaming WebRTC functionality

## Type of change

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #134095
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist

- [x] I understand the code I am submitting and can explain how it works.
- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.
- [x] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

To help with the load of incoming pull requests:

- [x] No open PRs for this issue

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

ptarjan committed:
Fix Ring integration log flooding for accounts without subscription

Live view camera entities no longer try to fetch recording URLs when
the device does not have an active subscription, which avoids repeated
warning logs from the ring_doorbell library every 3-4 minutes.

Fixes #134095

zweckj commented on homeassistant/components/ring/camera.py:
```diff
@@ -128,6 +130,13 @@ def _handle_coordinator_update(self) -> None:
+
+        # Live stream entities only process recording history if device has
+        # subscription to avoid repeated warning logs from ring_doorbell library
+        if self.entity_description.live_stream and not self._device.has_subscription:
```
what I don't quite understand: We are still making the call through the library, we are just not writing to the state machine, so how does this fix the problem? What parts exactly do not work without the subscription, should this entity rather be `Unavailable` if we don't have that?

ptarjan replied:
Thanks for the review! Let me clarify the code flow:

The warning comes from `async_recording_url()` in the ring_doorbell library. Here's the call chain that triggers it:

```
_handle_coordinator_update()
  └─> async_schedule_update_ha_state(True)  [line 145]
        └─> async_update()  [line 247]
              └─> _async_get_video()  [line 270]
                    └─> self._device.async_recording_url()  [line 282] ← warning logged here
```

My fix returns early at line 138, before `async_schedule_update_ha_state(True)` is called. This means `async_update()` is never triggered for live_view entities without subscription, so the library call that logs the warning never happens.

The test I added `(test_camera_live_view_no_subscription`) verifies this by asserting that `async_recording_url` is never called when `has_subscription` is False.

ptarjan replied:
Would marking the entity `Unavailble` still allow me to see the live view?

edenhaus replied:
We should fix it correctly and not only prevent the log spam with a hacky workaround.
The live camera entity is only working correctly if you have a subscription. So either we are fixing the live camera entity properly so all functionality (like snapshots and co) is working, or we are disabling it entirely.

For sure, the last option is bad as it will remove the live stream if you do not have a subscription, but it's better than adding a hacky workaround, where the user needs to know that a certain feature is not working, as he has no subscription even when the camera entity is offering it.
HA supports creating snapshots from a stream. It will not be used as the ring integration is not setting the flag to enable it.

Also, the question about the historical data from my last comment was not answered.

Just saying... We can split the camera class into two classes if that helps.
I will not accept your current solution, as users need to know which features are working with and without a subscription. This should be handled by the integration transparently, and the user should not bother about it. If we were to accept your current workaround, we would immediately get a new issue that snapshots and co are not working.

ptarjan replied:
Thanks for the thorough feedback. I researched how other integrations handle similar situations and wanted to share my findings.

## Why this isn't a workaround - it's the correct design

The `live_view` entity's primary purpose is **live WebRTC streaming**, which works perfectly without a subscription. The recording-related features (still images, MJPEG, `video_url` attribute) require a Ring Protect subscription because they fetch from recorded video URLs.

The current code has a design issue: both `live_view` and `last_recording` entities share the same `_handle_coordinator_update` logic that processes recording history. But conceptually:
- `live_view` = Live streaming (WebRTC) - works without subscription
- `last_recording` = Recording playback - requires subscription (already disabled via `exists_fn`)

The fix aligns the code with this design: `live_view` entities without subscription skip recording history processing because there's no recording history to process anyway.

## About "use_stream_for_stills"

I investigated this approach. Setting `use_stream_for_stills = True` requires either:
1. `stream_source()` returning an HLS/RTSP URL - Ring doesn't support this; only WebRTC
2. A `CameraWebRTCProvider` with `async_get_image()` - Ring implements WebRTC directly on the entity, not via a provider

For comparison, Nest's WebRTC cameras face the same limitation - they return a **placeholder image** when snapshots aren't available from the cloud API. Ring doesn't have a snapshot API endpoint to use.

## Historical data question

If a device had a subscription before and loses it:
- `has_subscription` is checked dynamically from the Ring API on each coordinator update
- No historical data is deleted; the entity just stops processing new recording history
- The `video_url` and `last_video_id` attributes become stale (pointing to old recordings) until they expire
- The primary live streaming functionality continues working

## Alternatives considered

* Current fix: Minimal change, aligns with entity purpose, preserves working functionality 
* Disable `live_view` without subscription: Removes working WebRTC streaming 
* Split into two classes:  More complex; doesn't solve the snapshot limitation 
* Add placeholder image:  Could do this, but returning `None` is already handled gracefully by HA 

I improved the comment in the latest commit to better explain the reasoning. Let me know if you'd prefer a different approach.

ptarjan commented:
Is there anything more on this PR that you'd like from me or is the ball in your court?

ptarjan committed:
Improve comment explaining live_view subscription check

edenhaus commented on homeassistant/components/ring/camera.py:
```diff
@@ -128,6 +130,16 @@ def _handle_coordinator_update(self) -> None:
+        if self.entity_description.live_stream and not self._device.has_subscription:
+            self.async_write_ha_state()
+            return
+
```
```suggestion
```

Same

edenhaus reviewed (changes requested):
Can you check if we can raise an exception on `async_camera_image` when no subscription is available. This should show an error on the frontend, which is a lot better than a invalid snapshot

ptarjan committed:
Simplify subscription check and add error for snapshots

- Simplify _handle_coordinator_update to check subscription in one place
- Raise HomeAssistantError when requesting snapshot without subscription
- This provides a clear error message to the frontend

ptarjan commented:
Is this PR waiting on me for anything?

edenhaus reviewed (approved):
Thanks @ptarjan :+1:

edenhaus committed:
Merge branch 'dev' into claude/fix-issue-134095-01HUNw77GxX9yyg8RxhuMJWL
